Design/Workbench/design-projectsaving.md

```diff
@@ -0,0 +1,229 @@
+
+Options for Implementation of Workspace Saving
+----------------------------------------------
+**Python:**
+- Pickling
+	- Pros: Built in library of Python
+	- Cons: Need to add _ _ getstate_ _ and _ _ setstate_ _ functions to every UI class which must return a pickle-able type (dict, list, int, string etc). Then setstate function. Unconfirmed if possible or not.
+- JSON Encoding
+	- Pros: Built in library of Python, "faster" than Pickling: https://www.benfrederickson.com/dont-pickle-your-data/, Potential to be integrated with C++ as it's universal
+	- Cons: Needs methods in every class, doesn't integrate fully with C++ (It won't translate a C++ class into a JSON file/string without an extra library)
+
+**C++**
+- Serialization with boost 
+	- Pros: We already have it, simple, uses an archive type for template serialization, has forward compatibility when classes change.
+	- Cons: We need to give Friend access to each UI class for the boost::serialize::access, not python.
+- Serialization with cereal
+	- Pros: Very easy to implement using a template of Archive type.
+	- Cons: We don't already have the cereal library, plus we'd need to cite it, as per licensing, not python.
+
+**Both C++ and Python**
+- Our own method of serialization utilizing a JSON back-end
+	- Pros: We can pass C++ values to Python based serializer and back to C++, we can determine file format, we can write custom conversions for types (Most Qt types have a QVariant type conversion to JSON anyway), highly flexible
+	- Cons: Requires extra work to implement that a already available library 
+
+
+Implementation of save window, the recommended Method
+-----------------------------------------------------
+Utilizing the serializing method with a JSON back-end, will likely be best as not only will it use a standard file type, it should be easy to change the file type by changing a few functions inside the serializer. Each UI that is serializable should have a encode and decode ability where encode would create a dictionary with a string as the key and the value set to whatever object you would need to save. Decode would use the string key to get back encoder's value and emplace the original value again, and update the GUI to represent these values.
+
+**Saving:**
+Utlizing an abstraction and encapsulation method we have a few major sections to tackle with a middle man "Saver Class". To save we would have the section which encompasses encoding. These sections are encoding, clumping encoded pieces together ("Saver") and writing, similar to how MVP works with GUIs. This way the encoding and writing parts will not need to touch eachother and can be swapped out independently as well as tested seperately with mocking and unit tests.
```

Comment:
The encoder/decoder classes seem to come a bit out of nowhere here. I think moving the sequence diagram above will help.

Reply:
Done :+1: